lib/roken/rtbl.3
.\" Copyright (c) 2004 Kungliga Tekniska Högskolan
.\" (Royal Institute of Technology, Stockholm, Sweden).
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" 3. Neither the name of the Institute nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\" $Id$
.\"
.Dd June 26, 2004
.Dt RTBL 3
.Os HEIMDAL
.Sh NAME
.Nm rtbl_create ,
.Nm rtbl_destroy ,
.Nm rtbl_set_flags ,
.Nm rtbl_get_flags ,
.Nm rtbl_set_prefix ,
.Nm rtbl_set_separator ,
.Nm rtbl_set_column_prefix ,
.Nm rtbl_set_column_affix_by_id ,
.Nm rtbl_add_column ,
.Nm rtbl_add_column_by_id ,
.Nm rtbl_add_column_entry ,
.Nm rtbl_add_column_entry_by_id ,
.Nm rtbl_new_row ,
.Nm rtbl_format
.Nd format data in simple tables
.Sh LIBRARY
The roken library (libroken, -lroken)
.Sh SYNOPSIS
.In rtbl.h
.Ft int
.Fn rtbl_add_column "rtbl_t table" "const char *column_name" "unsigned int flags"
.Ft int
.Fn rtbl_add_column_by_id "rtbl_t table" "unsigned int column_id" "const char *column_header" "unsigned int flags"
.Ft int
.Fn rtbl_add_column_entry "rtbl_t table" "const char *column_name" "const char *cell_entry"
.Ft int
.Fn rtbl_add_column_entry_by_id "rtbl_t table" "unsigned int column_id" "const char *cell_entry"
.Ft rtbl_t
.Fn rtbl_create "void"
.Ft void
.Fn rtbl_destroy "rtbl_t table"
.Ft int
.Fn rtbl_new_row "rtbl_t table"
.Ft int
.Fn rtbl_set_column_affix_by_id "rtbl_t table" "unsigned int column_id "const char *prefix" "const char *suffix"
.Ft int
.Fn rtbl_set_column_prefix "rtbl_t table" "const char *column_name" "const char *prefix"
.Ft "unsigned int"
.Fn rtbl_get_flags "rtbl_t table"
.Ft void
.Fn rtbl_set_flags "rtbl_t table" "unsigned int flags"
.Ft int
.Fn rtbl_set_prefix "rtbl_t table" "const char *prefix"
.Ft int
.Fn rtbl_set_separator "rtbl_t table" "const char *separator"
.Ft int
.Fn rtbl_format "rtbl_t table "FILE *file"
.Sh DESCRIPTION
This set of functions assemble a simple table consisting of rows and
columns, allowing it to be printed with certain options. Typical use
would be output from tools such as
.Xr ls 1
or
.Xr netstat 1 ,
where you have a fixed number of columns, but don't know the column
widthds before hand.
.Pp
A table is created with
.Fn rtbl_create
and destroyed with
.Fn rtbl_destroy .
.Pp
Global flags on the table are set with
.Fa rtbl_set_flags
and retrieved with
.Fa rtbl_get_flags .
At present the only defined flag is
.Dv RTBL_HEADER_STYLE_NONE
which suppresses printing the header.
.Pp
Before adding data to the table, one or more columns need to be
created. This would normally be done with
.Fn rtbl_add_column_by_id ,
.Fa column_id
is any number of your choice (it's used only to identify columns),
.Fa column_header
is the header to print at the top of the column, and
.Fa flags
are flags specific to this column. Currently the only defined flag is
.Dv RTBL_ALIGN_RIGHT ,
aligning column entries to the right. Columns are printed in the order
they are added.
.Pp
There's also a way to add columns by column name with
.Fn rtbl_add_column ,
but this is less flexible (you need unique header names), and is
considered deprecated.
.Pp
To add data to a column you use
.Fn rtbl_add_column_entry_by_id ,
where the
.Fa column_id
is the same as when the column was added (adding data to a
non-existent column is undefined), and
.Fa cell_entry
is whatever string you wish to include in that cell. It should not
include newlines.
For columns added with
.Fn rtbl_add_column
you must use
.Fn rtbl_add_column_entry
instead.
.Pp
.Fn rtbl_new_row
fills all columns with blank entries until they all have the same
number of rows.
.Pp
Each column can have a separate prefix and suffix, set with
.Fa rtbl_set_column_affix_by_id ;
.Fa rtbl_set_column_prefix
allows setting the prefix only by column name. In addition to this,
columns may be separated by a string set with
.Fa rtbl_set_separator ( Ns
by default columns are not seprated by anything).
.Pp
The finished table is printed to
.Fa file
with
.Fa rtbl_format .
.Sh EXAMPLES
This program:
.Bd -literal -offset xxxx
#include <stdio.h>
#include <rtbl.h>
int
main(int argc, char **argv)
{
    rtbl_t table;
    table = rtbl_create();
    rtbl_set_separator(table, "  ");
    rtbl_add_column_by_id(table, 0, "Column A", 0);
    rtbl_add_column_by_id(table, 1, "Column B", RTBL_ALIGN_RIGHT);
    rtbl_add_column_by_id(table, 2, "Column C", 0);
    rtbl_add_column_entry_by_id(table, 0, "A-1");
    rtbl_add_column_entry_by_id(table, 0, "A-2");
    rtbl_add_column_entry_by_id(table, 0, "A-3");
    rtbl_add_column_entry_by_id(table, 1, "B-1");
    rtbl_add_column_entry_by_id(table, 2, "C-1");
    rtbl_add_column_entry_by_id(table, 2, "C-2");
    rtbl_add_column_entry_by_id(table, 1, "B-2");
    rtbl_add_column_entry_by_id(table, 1, "B-3");
    rtbl_add_column_entry_by_id(table, 2, "C-3");
    rtbl_add_column_entry_by_id(table, 0, "A-4");
    rtbl_new_row(table);
    rtbl_add_column_entry_by_id(table, 1, "B-4");
    rtbl_new_row(table);
    rtbl_add_column_entry_by_id(table, 2, "C-4");
    rtbl_new_row(table);
    rtbl_format(table, stdout);
    rtbl_destroy(table);
    return 0;
}
.Ed
.Pp
will output the following:
.Bd -literal -offset xxxx
Column A  Column B  Column C
A-1            B-1  C-1
A-2            B-2  C-2
A-3            B-3  C-3
A-4
               B-4
                    C-4
.Ed
.\" .Sh SEE ALSO


NEWS
Release Notes - Heimdal - Version Heimdal 1.2.1

* Bug

  [HEIMDAL-147] - Heimdal 1.2 not compiling on Solaris
  [HEIMDAL-151] - Make canned tests work again after cert expired
  [HEIMDAL-152] - iprop test: use full hostname to avoid realm
                  resolving errors
  [HEIMDAL-153] - ftp: Use the correct length for unmap, msync

Release Notes - Heimdal - Version Heimdal 1.2

* Bug

  [HEIMDAL-10] - Follow-up on bug report for SEGFAULT in
  		 gss_display_name/gss_export_name when using SPNEGO
  [HEIMDAL-15] - Re: [Heimdal-bugs] potential bug in Heimdal 1.1
  [HEIMDAL-17] - Remove support for depricated [libdefaults]capath
  [HEIMDAL-52] - hdb overwrite aliases for db databases
  [HEIMDAL-54] - Two issues which affect credentials delegation
  [HEIMDAL-58] - sockbuf.c calls setsockopt with bad args
  [HEIMDAL-62] - Fix printing of sig_atomic_t
  [HEIMDAL-87] - heimdal 1.1 not building under cygwin in hcrypto
  [HEIMDAL-105] - rcp: sync rcp with upstream bsd rcp codebase
  [HEIMDAL-117] - Use libtool to detect symbol versioning (Debian Bug#453241)

* Improvement
  [HEIMDAL-67] - Fix locking and store credential in atomic writes
                 in the FILE credential cache
  [HEIMDAL-106] - make compile on cygwin again
  [HEIMDAL-107] - Replace old random key generation in des module
                  and use it with RAND_ function instead
  [HEIMDAL-115] - Better documentation and compatibility in hcrypto
                  in regards to OpenSSL

* New Feature
  [HEIMDAL-3] - pkinit alg agility PRF test vectors
  [HEIMDAL-14] - Add libwind to Heimdal
  [HEIMDAL-16] - Use libwind in hx509
  [HEIMDAL-55] - Add flag to krb5 to not add GSS-API INT|CONF to
                 the negotiation
  [HEIMDAL-74] - Add support to report extended error message back
                 in AS-REQ to support windows clients
  [HEIMDAL-116] - test pty based application (using rkpty)
  [HEIMDAL-120] - Use new OpenLDAP API (older deprecated)

* Task
  [HEIMDAL-63] - Dont try key usage KRB5_KU_AP_REQ_AUTH for TGS-REQ.
                 This drop compatibility with pre 0.3d KDCs.
  [HEIMDAL-64] - kcm: first implementation of kcm-move-cache
  [HEIMDAL-65] - Failed to compile with --disable-pk-init
  [HEIMDAL-80] - verify that [VU#162289]: gcc silently discards some
                 wraparound checks doesn't apply to Heimdal

Changes in release 1.1

 * Read-only PKCS11 provider built-in to hx509.

 * Documentation for hx509, hcrypto and ntlm libraries improved.

 * Better compatibilty with Windows 2008 Server pre-releases and Vista.

 * Mac OS X 10.5 support for native credential cache.

 * Provide pkg-config file for Heimdal (heimdal-gssapi.pc).

 * Bug fixes.

Changes in release 1.0.2

* Ubuntu packages.

* Bug fixes.

Changes in release 1.0.1

 * Serveral bug fixes to iprop.

 * Make work on platforms without dlopen.

 * Add RFC3526 modp group14 as default.

 * Handle [kdc] database = { } entries without realm = stanzas.

 * Make krb5_get_renewed_creds work.

 * Make kaserver preauth work again.

 * Bug fixes.

Changes in release 1.0

 * Add gss_pseudo_random() for mechglue and krb5.

 * Make session key for the krbtgt be selected by the best encryption
   type of the client.

 * Better interoperability with other PK-INIT implementations.

 * Inital support for Mac OS X Keychain for hx509.

 * Alias support for inital ticket requests.

 * Add symbol versioning to selected libraries on platforms that uses
   GNU link editor: gssapi, hcrypto, heimntlm, hx509, krb5, and libkdc.

 * New version of imath included in hcrypto.

 * Fix memory leaks.

 * Bugs fixes.

Changes in release 0.8.1

 * Make ASN.1 library less paranoid to with regard to NUL in string to
   make it inter-operate with MIT Kerberos again.

 * Make GSS-API library work again when using gss_acquire_cred

 * Add symbol versioning to libgssapi when using GNU ld.

 * Fix memory leaks 

 * Bugs fixes

Changes in release 0.8

 * PK-INIT support.

 * HDB extensions support, used by PK-INIT.

 * New ASN.1 compiler.

 * GSS-API mechglue from FreeBSD.

 * Updated SPNEGO to support RFC4178.

 * Support for Cryptosystem Negotiation Extension (RFC 4537).

 * A new X.509 library (hx509) and related crypto functions.

 * A new ntlm library (heimntlm) and related crypto functions.

 * Updated the built-in crypto library with bignum support using
   imath, support for RSA and DH and renamed it to libhcrypto.

 * Subsystem in the KDC, digest, that will perform the digest
   operation in the KDC, currently supports: CHAP, MS-CHAP-V2, SASL
   DIGEST-MD5 NTLMv1 and NTLMv2.

 * KDC will return the "response too big" error to force TCP retries
   for large (default 1400 bytes) UDP replies.  This is common for
   PK-INIT requests.

 * Libkafs defaults to use 2b tokens.

 * Default to use the API cache on Mac OS X.

 * krb5_kuserok() also checks ~/.k5login.d directory for acl files,
   see manpage for krb5_kuserok for description.

 * Many, many, other updates to code and info manual and manual pages.

 * Bug fixes

Changes in release 0.7.2

* Fix security problem in rshd that enable an attacker to overwrite
  and change ownership of any file that root could write.

* Fix a DOS in telnetd. The attacker could force the server to crash
  in a NULL de-reference before the user logged in, resulting in inetd
  turning telnetd off because it forked too fast.

* Make gss_acquire_cred(GSS_C_ACCEPT) check that the requested name
  exists in the keytab before returning success. This allows servers
  to check if its even possible to use GSSAPI.

* Fix receiving end of token delegation for GSS-API. It still wrongly
  uses subkey for sending for compatibility reasons, this will change
  in 0.8.

* telnetd, login and rshd are now more verbose in logging failed and
  successful logins.

* Bug fixes

Changes in release 0.7.1

* Bug fixes

Changes in release 0.7

 * Support for KCM, a process based credential cache

 * Support CCAPI credential cache

 * SPNEGO support


lib/roken/rtbl.c
/*
 * Copyright (c) 2000, 2002, 2004 Kungliga Tekniska Högskolan
 * (Royal Institute of Technology, Stockholm, Sweden).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the Institute nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
RCSID ("$Id$");
#endif
#include "roken.h"
#include "rtbl.h"

struct column_entry {
    char *data;
};

struct column_data {
    char *header;
    char *prefix;
    int width;
    unsigned flags;
    size_t num_rows;
    struct column_entry *rows;
    unsigned int column_id;
    char *suffix;
};

struct rtbl_data {
    char *column_prefix;
    size_t num_columns;
    struct column_data **columns;
    unsigned int flags;
    char *column_separator;
};

rtbl_t ROKEN_LIB_FUNCTION
rtbl_create (void)
{
    return calloc (1, sizeof (struct rtbl_data));
}

void ROKEN_LIB_FUNCTION
rtbl_set_flags (rtbl_t table, unsigned int flags)
{
    table->flags = flags;
}

unsigned int ROKEN_LIB_FUNCTION
rtbl_get_flags (rtbl_t table)
{
    return table->flags;
}

static struct column_data *
rtbl_get_column_by_id (rtbl_t table, unsigned int id)
{
    int i;
    for(i = 0; i < table->num_columns; i++)
	if(table->columns[i]->column_id == id)
	    return table->columns[i];
    return NULL;
}

static struct column_data *
rtbl_get_column (rtbl_t table, const char *column)
{
    int i;
    for(i = 0; i < table->num_columns; i++)
	if(strcmp(table->columns[i]->header, column) == 0)
	    return table->columns[i];
    return NULL;
}

void ROKEN_LIB_FUNCTION
rtbl_destroy (rtbl_t table)
{
    int i, j;

    for (i = 0; i < table->num_columns; i++) {
	struct column_data *c = table->columns[i];

	for (j = 0; j < c->num_rows; j++)
	    free (c->rows[j].data);
	free (c->rows);
	free (c->header);
	free (c->prefix);
	free (c->suffix);
	free (c);
    }
    free (table->column_prefix);
    free (table->column_separator);
    free (table->columns);
    free (table);
}

int ROKEN_LIB_FUNCTION
rtbl_add_column_by_id (rtbl_t table, unsigned int id,
		       const char *header, unsigned int flags)
{
    struct column_data *col, **tmp;

    tmp = realloc (table->columns, (table->num_columns + 1) * sizeof (*tmp));
    if (tmp == NULL)
	return ENOMEM;
    table->columns = tmp;
    col = malloc (sizeof (*col));
    if (col == NULL)
	return ENOMEM;
    col->header = strdup (header);
    if (col->header == NULL) {
	free (col);
	return ENOMEM;
    }
    col->prefix = NULL;
    col->width = 0;
    col->flags = flags;
    col->num_rows = 0;
    col->rows = NULL;
    col->column_id = id;
    col->suffix = NULL;
    table->columns[table->num_columns++] = col;
    return 0;
}

int ROKEN_LIB_FUNCTION
rtbl_add_column (rtbl_t table, const char *header, unsigned int flags)
{
    return rtbl_add_column_by_id(table, 0, header, flags);
}

int ROKEN_LIB_FUNCTION
rtbl_new_row(rtbl_t table)
{
    size_t max_rows = 0;
    size_t c;
    for (c = 0; c < table->num_columns; c++)
	if(table->columns[c]->num_rows > max_rows)
	    max_rows = table->columns[c]->num_rows;
    for (c = 0; c < table->num_columns; c++) {
	struct column_entry *tmp;

	if(table->columns[c]->num_rows == max_rows)
	    continue;
	tmp = realloc(table->columns[c]->rows,
		      max_rows * sizeof(table->columns[c]->rows));
	if(tmp == NULL)
	    return ENOMEM;
	table->columns[c]->rows = tmp;
	while(table->columns[c]->num_rows < max_rows) {
	    if((tmp[table->columns[c]->num_rows++].data = strdup("")) == NULL)
		return ENOMEM;
	}
    }
    return 0;
}

static void
column_compute_width (rtbl_t table, struct column_data *column)
{
    int i;

    if(table->flags & RTBL_HEADER_STYLE_NONE)
	column->width = 0;
    else
	column->width = strlen (column->header);
    for (i = 0; i < column->num_rows; i++)
	column->width = max (column->width, strlen (column->rows[i].data));
}

/* DEPRECATED */
int ROKEN_LIB_FUNCTION
rtbl_set_prefix (rtbl_t table, const char *prefix)
{
    if (table->column_prefix)
	free (table->column_prefix);
    table->column_prefix = strdup (prefix);
    if (table->column_prefix == NULL)
	return ENOMEM;
    return 0;
}

int ROKEN_LIB_FUNCTION
rtbl_set_separator (rtbl_t table, const char *separator)
{
    if (table->column_separator)
	free (table->column_separator);
    table->column_separator = strdup (separator);
    if (table->column_separator == NULL)
	return ENOMEM;
    return 0;
}

int ROKEN_LIB_FUNCTION
rtbl_set_column_prefix (rtbl_t table, const char *column,
			const char *prefix)
{
    struct column_data *c = rtbl_get_column (table, column);

    if (c == NULL)
	return -1;
    if (c->prefix)
	free (c->prefix);
    c->prefix = strdup (prefix);
    if (c->prefix == NULL)
	return ENOMEM;
    return 0;
}

int ROKEN_LIB_FUNCTION
rtbl_set_column_affix_by_id(rtbl_t table, unsigned int id,
			    const char *prefix, const char *suffix)
{
    struct column_data *c = rtbl_get_column_by_id (table, id);

    if (c == NULL)
	return -1;
    if (c->prefix)
	free (c->prefix);
    if(prefix == NULL)
	c->prefix = NULL;
    else {
	c->prefix = strdup (prefix);
	if (c->prefix == NULL)
	    return ENOMEM;
    }

    if (c->suffix)
	free (c->suffix);
    if(suffix == NULL)
	c->suffix = NULL;
    else {
	c->suffix = strdup (suffix);
	if (c->suffix == NULL)
	    return ENOMEM;
    }
    return 0;
}


static const char *
get_column_prefix (rtbl_t table, struct column_data *c)
{
    if (c == NULL)
	return "";
    if (c->prefix)
	return c->prefix;
    if (table->column_prefix)
	return table->column_prefix;
    return "";
}

static const char *
get_column_suffix (rtbl_t table, struct column_data *c)
{
    if (c && c->suffix)
	return c->suffix;
    return "";
}

static int
add_column_entry (struct column_data *c, const char *data)
{
    struct column_entry row, *tmp;

    row.data = strdup (data);
    if (row.data == NULL)
	return ENOMEM;
    tmp = realloc (c->rows, (c->num_rows + 1) * sizeof (*tmp));
    if (tmp == NULL) {
	free (row.data);
	return ENOMEM;
    }
    c->rows = tmp;
    c->rows[c->num_rows++] = row;
    return 0;
}

int ROKEN_LIB_FUNCTION
rtbl_add_column_entry_by_id (rtbl_t table, unsigned int id, const char *data)
{
    struct column_data *c = rtbl_get_column_by_id (table, id);

    if (c == NULL)
	return -1;

    return add_column_entry(c, data);
}

int ROKEN_LIB_FUNCTION
rtbl_add_column_entryv_by_id (rtbl_t table, unsigned int id,
			      const char *fmt, ...)
{
    va_list ap;
    char *str;
    int ret;

    va_start(ap, fmt);
    ret = vasprintf(&str, fmt, ap);
    va_end(ap);
    if (ret == -1)
	return -1;
    ret = rtbl_add_column_entry_by_id(table, id, str);
    free(str);
    return ret;
}

int ROKEN_LIB_FUNCTION
rtbl_add_column_entry (rtbl_t table, const char *column, const char *data)
{
    struct column_data *c = rtbl_get_column (table, column);

    if (c == NULL)
	return -1;

    return add_column_entry(c, data);
}

int ROKEN_LIB_FUNCTION
rtbl_add_column_entryv (rtbl_t table, const char *column, const char *fmt, ...)
{
    va_list ap;
    char *str;
    int ret;

    va_start(ap, fmt);
    ret = vasprintf(&str, fmt, ap);
    va_end(ap);
    if (ret == -1)
	return -1;
    ret = rtbl_add_column_entry(table, column, str);
    free(str);
    return ret;
}


int ROKEN_LIB_FUNCTION
rtbl_format (rtbl_t table, FILE * f)
{
    int i, j;

    for (i = 0; i < table->num_columns; i++)
	column_compute_width (table, table->columns[i]);
    if((table->flags & RTBL_HEADER_STYLE_NONE) == 0) {
	for (i = 0; i < table->num_columns; i++) {
	    struct column_data *c = table->columns[i];

	    if(table->column_separator != NULL && i > 0)
		fprintf (f, "%s", table->column_separator);
	    fprintf (f, "%s", get_column_prefix (table, c));
	    if(i == table->num_columns - 1 && c->suffix == NULL)
		/* last column, so no need to pad with spaces */
		fprintf (f, "%-*s", 0, c->header);
	    else
		fprintf (f, "%-*s", (int)c->width, c->header);
	    fprintf (f, "%s", get_column_suffix (table, c));
	}
	fprintf (f, "\n");
    }

    for (j = 0;; j++) {
	int flag = 0;

	/* are there any more rows left? */
	for (i = 0; flag == 0 && i < table->num_columns; ++i) {
	    struct column_data *c = table->columns[i];

	    if (c->num_rows > j) {
		++flag;
		break;
	    }
	}
	if (flag == 0)
	    break;

	for (i = 0; i < table->num_columns; i++) {
	    int w;
	    struct column_data *c = table->columns[i];

	    if(table->column_separator != NULL && i > 0)
		fprintf (f, "%s", table->column_separator);

	    w = c->width;

	    if ((c->flags & RTBL_ALIGN_RIGHT) == 0) {
		if(i == table->num_columns - 1 && c->suffix == NULL)
		    /* last column, so no need to pad with spaces */
		    w = 0;
		else
		    w = -w;
	    }
	    fprintf (f, "%s", get_column_prefix (table, c));
	    if (c->num_rows <= j)
		fprintf (f, "%*s", w, "");
	    else
		fprintf (f, "%*s", w, c->rows[j].data);
	    fprintf (f, "%s", get_column_suffix (table, c));
	}
	fprintf (f, "\n");
    }
    return 0;
}

#ifdef TEST
int
main (int argc, char **argv)
{
    rtbl_t table;

    table = rtbl_create ();
    rtbl_add_column_by_id (table, 0, "Issued", 0);
    rtbl_add_column_by_id (table, 1, "Expires", 0);
    rtbl_add_column_by_id (table, 2, "Foo", RTBL_ALIGN_RIGHT);
    rtbl_add_column_by_id (table, 3, "Principal", 0);

    rtbl_add_column_entry_by_id (table, 0, "Jul  7 21:19:29");
    rtbl_add_column_entry_by_id (table, 1, "Jul  8 07:19:29");
    rtbl_add_column_entry_by_id (table, 2, "73");
    rtbl_add_column_entry_by_id (table, 2, "0");
    rtbl_add_column_entry_by_id (table, 2, "-2000");
    rtbl_add_column_entry_by_id (table, 3, "krbtgt/NADA.KTH.SE@NADA.KTH.SE");

    rtbl_add_column_entry_by_id (table, 0, "Jul  7 21:19:29");
    rtbl_add_column_entry_by_id (table, 1, "Jul  8 07:19:29");
    rtbl_add_column_entry_by_id (table, 3, "afs/pdc.kth.se@NADA.KTH.SE");

    rtbl_add_column_entry_by_id (table, 0, "Jul  7 21:19:29");
    rtbl_add_column_entry_by_id (table, 1, "Jul  8 07:19:29");
    rtbl_add_column_entry_by_id (table, 3, "afs@NADA.KTH.SE");

    rtbl_set_separator (table, "  ");

    rtbl_format (table, stdout);

    rtbl_destroy (table);

    printf("\n");

    table = rtbl_create ();
    rtbl_add_column_by_id (table, 0, "Column A", 0);
    rtbl_set_column_affix_by_id (table, 0, "<", ">");
    rtbl_add_column_by_id (table, 1, "Column B", 0);
    rtbl_set_column_affix_by_id (table, 1, "[", "]");
    rtbl_add_column_by_id (table, 2, "Column C", 0);
    rtbl_set_column_affix_by_id (table, 2, "(", ")");

    rtbl_add_column_entry_by_id (table, 0, "1");
    rtbl_new_row(table);
    rtbl_add_column_entry_by_id (table, 1, "2");
    rtbl_new_row(table);
    rtbl_add_column_entry_by_id (table, 2, "3");
    rtbl_new_row(table);

    rtbl_set_separator (table, "  ");
    rtbl_format (table, stdout);

    rtbl_destroy (table);

    return 0;
}

#endif
